src/gc/pooltable.d

```diff
@@ -0,0 +1,287 @@
+/**
+ * A sorted array to quickly lookup pools.
+ *
+ * Copyright: Copyright Digital Mars 2001 -.
+ * License:   $(WEB www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+ * Authors:   Walter Bright, David Friedman, Sean Kelly, Martin Nowak
+ */
+module gc.pooltable;
+
+static import cstdlib=core.stdc.stdlib;
+
+struct PoolTable(Pool)
+{
+    import core.stdc.string : memmove;
+
+nothrow:
+    void reset()
+    {
+        cstdlib.free(pools);
+        pools = null;
+        npools = 0;
+    }
+
+    bool insert(Pool* pool)
+    {
+        auto newpools = cast(Pool **)cstdlib.realloc(pools, (npools + 1) * pools[0].sizeof);
```

Comment:
> Freeing pools? Sure, happens whenever your men usage goes low.

It still needs all objects/pages in the pool to be unused. This might happen for large object pools, but small objects are more or less randomly distributed over the pools. It only needs a few objects to pin all pools (with the default settings an application with 200 MB pool memory has only 11 pools IIRC).